Prime inventory if FRU’s VPD file doesn’t exists.

branupama opened the pull request:
While collecting FRU’s VPD data, if the EEPROM file path is not exists for that FRU then prime inventory for that FRU.
For all the inventory paths under the FRU’s EEPROM from JSON config file, priming will populate interface property map and calls PIM to publish populated data on DBUs.

The following information for each inventory paths will be published under corresponding Object path’s.
    1.	Property map for commonInterfaces.
    2.	Property map for extraInterfaces.
    3.	Present property.

Testing:
By adding extra logs, observed ObjectMap(prime data) is created for FRU’s which EEPROM path is not accessible when vpd-manger application runs.

branupama commented on src/utils.cpp:
```diff
@@ -281,125 +281,124 @@ std::string getExpandedLocationCode(const std::string& unexpandedLocationCode,
-                                 "chassis/motherboard",
-                                 interfaceList);
+        std::string firstKwdValue, secondKwdValue;
+        types::IPZVpdMap::const_iterator itrToVCEN;
```
I have  removed top if case and only added this if case for getExpandedLocationCode API.
all these changes appearing because of the clang alignment happened.

SunnySrivastava1984 replied:
Why this change was required, please mention that as well.

branupama replied:
I initially thought getting IPZVpdMap type from an empty VPDMapVariant and dereferencing an empty map will cause an issue, so bought this check inside to get location code from DBus in case priming workflow.
I have checked the code again and find out that this change is not required.
I will revert back this change.

SunnySrivastava1984 replied:
Ok please update along with another comments.

SunnySrivastava1984 commented on src/worker.cpp:
```diff
@@ -653,12 +658,44 @@ bool Worker::isCPUIOGoodOnly(const std::string& pgKeyword)
 
-void Worker::primeInventory(const types::IPZVpdMap& ipzVpdMap,
-                            types::ObjectMap primeObjects)
+void Worker::primeInventory(const std::string i_vpdFilePath,
```
Check for valid parameters.

branupama replied:
Updated.

SunnySrivastava1984 commented on src/worker.cpp:
```diff
@@ -653,12 +658,44 @@ bool Worker::isCPUIOGoodOnly(const std::string& pgKeyword)
+            io_objectInterfaceMap.emplace(std::move(l_fruObjectPath),
+                                          std::move(l_interfaces));
+        }
+    }
```
If the JSON is empty then what happens?

branupama replied:
Updated the code .

SunnySrivastava1984 commented on src/worker.cpp:
```diff
@@ -860,11 +897,8 @@ void Worker::publishSystemVPD(const types::VPDMapVariant& parsedVpdMap)
 
     if (auto ipzVpdMap = std::get_if<types::IPZVpdMap>(&parsedVpdMap))
     {
+        (void)ipzVpdMap;
```
Why this is needed here?

branupama replied:
for ipzVpdMap variable I was getting 'unused variable' error when I compile locally using ninja on docker.
as ipzVpdMap variable is not being used inside if case, I have removed creating this variable.

SunnySrivastava1984 commented on src/worker.cpp:
```diff
@@ -602,6 +603,10 @@ void Worker::populateInterfaces(const nlohmann::json& interfaceJson,
                     }
+                    else
+                    {
+                        propertyMap.emplace(property, "");
```
Why this is required?

SunnySrivastava1984 replied:
Why empty value should be passed? What will happen is nothing is passed?

branupama replied:
this is the place where handling value of Object type from JSON, ex: extracting common interfaces.
If ipzVpdMap map is not empty, we are getting value from ipzVpdMap.
if ipzVpdMap is empty I wanted to create property for common interfaces while priming inventory with property(like PartNumber, SerialNumber ..) and empty string value.

SunnySrivastava1984 replied:
Ok, Can we have a demo of this workflow? Would like to see working of PIM in these scenarios.

branupama replied:
sure.

branupama replied:
Removed the above changes, as static properties updated in PIM by default.

SunnySrivastava1984 commented on include/worker.hpp:
```diff
@@ -300,11 +300,11 @@ class Worker
      */
-    void primeInventory(const types::IPZVpdMap& ipzVpdMap,
-                        types::ObjectMap primeObjects);
+    void primeInventory(const std::string i_vpdFilePath,
```
Use reference.

SunnySrivastava1984 commented on src/worker.cpp:
```diff
@@ -651,12 +656,56 @@ bool Worker::isCPUIOGoodOnly(const std::string& pgKeyword)
+        throw std::runtime_error("Empty VPD file path given");
+    }
+
+    if (!m_parsedJson.empty())
```
I would suggest check for json and throw from here. get rid of else part.
This will reduce the level of indentation and will make the code more readable.

branupama replied:
updated.

SunnySrivastava1984 commented on src/worker.cpp:
```diff
@@ -651,12 +656,56 @@ bool Worker::isCPUIOGoodOnly(const std::string& pgKeyword)
+        for (const auto& l_aFru : m_parsedJson["frus"][i_vpdFilePath])
+        {
+            types::InterfaceMap l_interfaces;
+            const auto& l_inventoryPath = l_aFru["inventoryPath"];
```
get rid of "l_inventoryPath" variable. Don't see any need.

SunnySrivastava1984 commented on src/worker.cpp:
```diff
@@ -968,6 +1013,15 @@ types::VPDMapVariant Worker::parseVpdFile(const std::string& i_vpdFilePath)
         }
 
+        types::ObjectMap l_objectInterfaceMap;
+        primeInventory(i_vpdFilePath, l_objectInterfaceMap);
```
Priming of inventory should always happen in case there is any failure in collection of the FRU.
In current situation, in case there is a throw from line 1010. I don't see "primeInventory" getting executed in that case. Please find another appropriate place to call your API.

branupama replied:
Both PrimeInventory and executePostFailAction should be called if i_vpdFilePath file doesn't exists.
even if I call primeInventory first and later executePostFailAction, it will impact executePostFailAction if there is any exception from primeInventory.
So another place we can call primeInventory is in parseAndPublishVPD catch block. 
But am not sure, is it correct to do so. @SunnySrivastava1984  please provide me your input here.

SunnySrivastava1984 replied:
> But am not sure

What are your doubts?

branupama replied:
Shall I call PrimeInventory from  parseAndPublishVPD's catch block ?

SunnySrivastava1984 replied:
That is the answer you need to come up with?

I just suggested a place from where you can call but at the end of the day, requirement is,
- In case there is any failure in the process of collecting a FRU then priming needs to be done for that FRU. 

So if you think that is the best place, go ahead, if not then see what can be other place(s) from where it can be called.

SunnySrivastava1984 commented on src/worker.cpp:
```diff
@@ -1020,8 +1063,17 @@ std::tuple<bool, std::string>
+        // Prime the inventry for FRUs which
         // are not present/processing had some error.
+        types::ObjectMap l_objectInterfaceMap;
+        primeInventory(i_vpdFilePath, l_objectInterfaceMap);
```
Calling it here can crash the service in some situations. Please fix that.

branupama replied:
Updated .

branupama replied:
I have mistakenly updated the comment here. 
Currently updated primeInventory API with success or failure status now 
and added log  to print if prime inventory fails .

SunnySrivastava1984 commented on src/worker.cpp:
```diff
@@ -1020,8 +1074,14 @@ std::tuple<bool, std::string>
-        // TODO: before returning, we should prime the inventry for FRUs which
+        // Prime the inventry for FRUs which
         // are not present/processing had some error.
+        if (!primeInventory(i_vpdFilePath))
```
It will still fail the service in some scenarios.
Please check.

SunnySrivastava1984 reviewed (changes requested):
Some changes required in commit message
 >While collecting FRU’s VPD data,

VPD itself has data, so VPD data is not correct.

> if the EEPROM file path is not exists

Does not exist.

> then prime inventory for that FRU.

Also add why.

>priming will populate interface property map and calls PIM to
publish populated data on DBUs.

Priming will populate map with empty values.

branupama commented on src/worker.cpp:
```diff
@@ -1020,8 +1070,14 @@ std::tuple<bool, std::string>
         // are not present/processing had some error.
+        if (!primeInventory(i_vpdFilePath))
+        {
+            logging::logMessage("Prime inventory is failed for FRU " +
```
Updated the log.

branupama commented:
> Some changes required in commit message
> 
> > While collecting FRU’s VPD data,
> 
> VPD itself has data, so VPD data is not correct.
> 
> > if the EEPROM file path is not exists
> 
> Does not exist.
> 
> > then prime inventory for that FRU.
> 
> Also add why.
> 
> > priming will populate interface property map and calls PIM to
> > publish populated data on DBUs.
> 
> Priming will populate map with empty values.

Updated the comment.

> Some changes required in commit message
> 
> > While collecting FRU’s VPD data,
> 
> VPD itself has data, so VPD data is not correct.
> 
> > if the EEPROM file path is not exists
> 
> Does not exist.
> 
> > then prime inventory for that FRU.
> 
> Also add why.
> 
> > priming will populate interface property map and calls PIM to
> > publish populated data on DBUs.
> 
> Priming will populate map with empty values.

Updated the comment.

SunnySrivastava1984 reviewed (approved):
Have some concern with the commit message though.
Rest looks ok to me.

SunnySrivastava1984 commented on src/worker.cpp:
```diff
@@ -651,12 +652,65 @@ bool Worker::isCPUIOGoodOnly(const std::string& pgKeyword)
+        if ((l_aFru.value("embedded", true)) &&
+            (!l_aFru.value("synthesized", false)))
+        {
+            processEmbeddedAndSynthesizedFrus(l_aFru, l_interfaces);
```
Going by the flow here, you are marking present as true. Where are you marking present as false?
Ideally, priming is done when FRU is not present.

SunnySrivastava1984 commented on src/worker.cpp:
```diff
@@ -651,12 +652,65 @@ bool Worker::isCPUIOGoodOnly(const std::string& pgKeyword)
+            m_parsedJson.contains("commonInterfaces"))
+        {
+            populateInterfaces(m_parsedJson["commonInterfaces"], l_interfaces,
+                               types::IPZVpdMap{});
```
Same here, get rid of having types attached with empty map. See if std::monotonic can be used here.

branupama replied:
Updated

branupama commented on src/worker.cpp:
```diff
@@ -651,12 +652,65 @@ bool Worker::isCPUIOGoodOnly(const std::string& pgKeyword)
+        if (l_aFru.contains("extraInterfaces"))
+        {
+            populateInterfaces(l_aFru["extraInterfaces"], l_interfaces,
+                               types::IPZVpdMap{});
```
I need to send empty types::IPZVpdMap here to get "LocationCode" from getExpandedLocationCode API .

branupama replied:
updated the code with std::monostate as default value, unexpanded location code from JSON config will be populated by default.

branupama commented on src/worker.cpp:
```diff
@@ -651,12 +652,65 @@ bool Worker::isCPUIOGoodOnly(const std::string& pgKeyword)
+    if (m_parsedJson.empty())
+    {
+        logging::logMessage(
+            "A config JSON is required to prime an inventory for a FRU " +
```
Updated.

branupama commented on src/worker.cpp:
```diff
@@ -651,12 +652,65 @@ bool Worker::isCPUIOGoodOnly(const std::string& pgKeyword)
+    }
+
+    types::ObjectMap l_objectInterfaceMap;
+    for (const auto& l_aFru : m_parsedJson["frus"][i_vpdFilePath])
```
Updated

SunnySrivastava1984 commented on src/worker.cpp:
```diff
@@ -651,12 +652,65 @@ bool Worker::isCPUIOGoodOnly(const std::string& pgKeyword)
+        types::InterfaceMap l_interfaces;
+        sdbusplus::message::object_path l_fruObjectPath(l_Fru["inventoryPath"]);
+
+        types::VPDMapVariant l_emptyValue;
```
Why is this required? Why don't you use empty value directly at calling point?

branupama replied:
populateInterfaces API is expecting reference variable, so can't send rvalue to it.

SunnySrivastava1984 replied:
So? You are saying that will a compile time error?

branupama replied:
yes, will get compile time error.

SunnySrivastava1984 replied:
Again, you know/checked or you are assuming?

branupama replied:
I try passing variable of types::VPDMapVariant  type directly to populateInterfaces(), got an error yesterday.
and I even tried with small test code today, got the error. 
Please find the error snippet,
test_ref.cpp:9:14: error: cannot bind non-const lvalue reference of type ‘int&’ to an rvalue of type ‘int’
    9 |         func(3);
      |              ^
test_ref.cpp:5:16: note:   initializing argument 1 of ‘void func(int&)’
    5 | void func(int &a){
      |           ~~~~~^

SunnySrivastava1984 replied:
So, the test case that you have put here and the actual scenario is same?

branupama replied:
yes, the above test scenario is missing const declaration, with const we can pass the rvalue.
updated the code, passing monostate directly now .

SunnySrivastava1984 commented on src/worker.cpp:
```diff
@@ -651,12 +652,65 @@ bool Worker::isCPUIOGoodOnly(const std::string& pgKeyword)
+                               l_emptyValue);
+        }
+
+        types::PropertyMap presProp;
```
What is "presProp". Please use meaningful names/abbreviations.

branupama replied:
updated.

SunnySrivastava1984 commented on src/worker.cpp:
```diff
@@ -651,12 +652,65 @@ bool Worker::isCPUIOGoodOnly(const std::string& pgKeyword)
+        {
+            presProp["Present"] = true;
+        }
+        utils::insertOrMerge(l_interfaces, "xyz.openbmc_project.Inventory.Item",
```
a blank line after line 691.

branupama replied:
Updated.

SunnySrivastava1984 reviewed (changes requested):
> if the EEPROM file path doesn’t exists

Also happens when collection fails for some reason.

>  waiting applications can consume it.

There are no waiting application. Till D-Bus is ok.

> priming will populate map with empty values.

Populate D-Bus with empty keyword value. Map is internal to implementation.

branupama commented:
> > if the EEPROM file path doesn’t exists
> 
> Also happens when collection fails for some reason.
> 
> > waiting applications can consume it.
> 
> There are no waiting application. Till D-Bus is ok.
> 
> > priming will populate map with empty values.
> 
> Populate D-Bus with empty keyword value. Map is internal to implementation.

Updated the commit message.

SunnySrivastava1984 commented:
> > Some changes required in commit message
> > > While collecting FRU’s VPD data,
> > 
> > 
> > VPD itself has data, so VPD data is not correct.
> > > if the EEPROM file path is not exists
> > 
> > 
> > Does not exist.
> > > then prime inventory for that FRU.
> > 
> > 
> > Also add why.
> > > priming will populate interface property map and calls PIM to
> > > publish populated data on DBUs.
> > 
> > 
> > Priming will populate map with empty values.
> 
> Updated the comment.
> 
> > Some changes required in commit message
> > > While collecting FRU’s VPD data,
> > 
> > 
> > VPD itself has data, so VPD data is not correct.
> > > if the EEPROM file path is not exists
> > 
> > 
> > Does not exist.
> > > then prime inventory for that FRU.
> > 
> > 
> > Also add why.
> > > priming will populate interface property map and calls PIM to
> > > publish populated data on DBUs.
> > 
> > 
> > Priming will populate map with empty values.
> 
> Updated the comment.

Priming will also take place when collection fails for reason other than file not present.

SunnySrivastava1984 reviewed (approved):
Looks good to me.
